### ISSUE TITLE
Decimal average accumulator type must have arbitrary precision

### ISSUE THREAD

**Issue by karteekmurthys** [facebookincubator/velox#2736]
In Presto JAVA worker, the decimal average aggregate functions uses `BigDecimal` type as an intermediate type. Since `BigDecimal` has arbitrary precision, the Total sum accumulated for a group/ global aggregation is allowed to cross the LONG DECIMAL limit of `10^38 - 1`. But, Velox has no such arbitrary type and is limited to use `UnscaledLongDecimal` as the accumulator type. This can be further improved if there is an `INT128` type to represent the accumulators, this adds more precision digits but not same as the Presto Java behavior.

Here is the PR that adds decimal average function: https://github.com/facebookincubator/velox/pull/2639
BigDecimal: https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html

We may have to implement a new VELOX Type similar to BigDecimal to achieve parity with Presto Java worker.

**Comment by karteekmurthys** [comment]
cc: @majetideepak

**Comment by majetideepak** [comment]
@Yuhta I wanted to move to iterative mean as well. However, it is computationally expensive (extra 1 X multiplication, 2 X division for each input). Will it be okay?

**Comment by majetideepak** [comment]
I guess BigDecimal arithmetic is also computationally expensive since it will involve handling multiple words.
I guess iterative mean makes sense for decimals.

**Comment by Yuhta** [comment]
@majetideepak We can avoid some divisions by keeping a partial sum, and only do the iterative averaging when the partial sum is about to overflow.  After that we can accumulate the partial sum again from zero.

**Comment by majetideepak** [comment]
@Yuhta You are right! We check for overflow for each input anyway. Your approach makes sense. Thanks.

**Comment by karteekmurthys** [comment]
Just to clarify, computing the average in parts may deviate from averages as a whole. This happens because we are performing UnscaleLongDecimal division which rounds up fraction to the nearest integer. This is less likely to happen in case of Presto Java worker where it stores intermediate results in BigDecimal.

**Comment by Yuhta** [comment]
@karteekmurthys We are using `double` for sum, not any integer type, so the division should not result in any rounding.  When you merge 2 averages you need to take the count into consideration as well and calculate the weighted average.

**Comment by Yuhta** [comment]
@karteekmurthys I see, in your PR the sum is kept in decimal type.  In this case we can keep track of the remainder as well to prevent rounding error.

**Comment by majetideepak** [comment]
@Yuhta great suggestion!

**Comment by karteekmurthys** [comment]
Thanks for the suggestions @Yuhta.

**Comment by karteekmurthys** [comment]
I have applied iterative mean technique in this PR: https://github.com/facebookincubator/velox/pull/2763

One caveat is that Iteratively computing average would not behave as expected while requesting for Partial and Intermediate aggregations. What is the impact of not having partial and intermediate aggregations in place..will it affect the overall average function? does the query plan ever request for only partial/intermediate aggregations..the user may not as there is no support for that in SQL AFAIK.

**Comment by Yuhta** [comment]
@karteekmurthys It should work with partial and intermediate aggregations, but we may need to change the intermediate type.  I will take a look at the PR these 2 days.

**Comment by karteekmurthys** [comment]
On Presto Java, the decimal average intermediate type is **not BigDecimal** but a **VARBINARY**. The final average step uses BigDecimal to compute total sum and average.

The intermediate result is wrapped in a VARBINARY and consists of **_count, overflow (if the sum so far overflown or underflown) and total sum._**.


**presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java**
```
private static BigDecimal average(long count, long overflow, Slice unscaledDecimal, int scale)
    {
        BigDecimal sum = new BigDecimal(Decimals.decodeUnscaledValue(unscaledDecimal), scale);
        if (overflow != 0) {
            sum = sum.add(new BigDecimal(OVERFLOW_MULTIPLIER.multiply(BigInteger.valueOf(overflow))));
        }
        return sum.divide(BigDecimal.valueOf(count), scale, ROUND_HALF_UP);
    }
```

**Comment by Yuhta** [comment]
@karteekmurthys Isn't `BigDecimal` in Java has arbitrary precision?  How come it overflows?

**Comment by karteekmurthys** [comment]
@Yuhta Sorry, I should have clarified earlier. The total sum in the intermediate state is a `Slice` of 16 bytes not BigDecimal.

Only the final aggregation uses BigDecimal.

**Comment by Yuhta** [comment]
@karteekmurthys So there is also an option for us to do the same thing as java to just sum in partial & intermediate aggregation.  We still need to handle the overflow in final aggregation though.

**Comment by karteekmurthys** [comment]
@Yuhta That is correct. I am working on simple test code on how to achieve this in Velox. We will follow similar approach of Java of tracking the overflow and just storing the sum in uint128_t.

The `sum = sum.add(new BigDecimal(OVERFLOW_MULTIPLIER.multiply(BigInteger.valueOf(overflow))));` step needs to be broken down into performing division by count followed by adding the divided parts.

**Comment by Yuhta** [comment]
@karteekmurthys It might not be easy in Velox.  Since you have to handle overflow in final aggregation anyway, it might worth just use the same implementation in both and give better result than Java.

**Comment by majetideepak** [comment]
This is resolved with LongDecimalWithOverflowState
